File: core/prompts/briefing.py
from core.prompts.voice import get_voice, BLOCKED_WORDS
from core.prompts.guards import inject_guards
# BriefingContext is imported lazily inside build_pulse_briefing_prompt to
# avoid a circular import (core.pulse.briefing imports this module, and
# core.pulse.models re-exports the pulse package).


def build_daily_brief_prompt(
    now_str: str,
    day_label: str,
    calendar_text: str,
    overdue_text: str,
    todo_text: str,
    recent_done_text: str,
    user_name: str | None = None,
) -> str:
    """Daily brief prompt. Used by dispatch.py handle_daily_brief.

    `user_name` comes from user_settings (fallback: env USER_NAME / "Danny").
    """
    from core.services.user_settings import resolve_user_name
    user_name = user_name or resolve_user_name()
    voice = get_voice()

    return f"""{voice}

CURRENT TIME: {now_str}

{user_name} wants their daily brief for {day_label}. You have their calendar, active tasks, overdue items, and recent completions. Identify what matters and cut through the noise.

Structure:
- Open with 1-2 sentences in Rhodey's voice: what's new, what's on top, what needs {user_name}'s attention today. This opening is ALWAYS required — never start with a section header or the calendar.
- Calendar events second. If an event is marked [PAST], note it already happened.
- **Context:** section third: 1-3 sentences on overdue items, blockers, urgency.
- Stop after context. No analysis of your own response.

Format:
- Emoji at the start of each line
- **bold** for section breaks (no ### headers)
- Bullet points only, no numbered lists
- Preserve [Project] brackets from task data exactly as shown

Example:
**Focus here** — bottleneck callout.
- 💰 Task name [Project]
- 📋 Another task [Project]

{day_label.upper()} — DATA CONTEXT:

CALENDAR EVENTS:
{calendar_text or "None"}

OVERDUE:
{overdue_text or "None"}

ACTIVE TASKS:
{todo_text or "None"}

RECENTLY COMPLETED (24h):
{recent_done_text or "None"}"""


def build_pulse_briefing_prompt(ctx, user_name: str | None = None, sections=None) -> str:
    """Build the pulse briefing prompt (M2/M9.3 de-personalized).

    `user_name` comes from user_settings (fallback: env USER_NAME / "Danny").
    `sections` is the per-tenant briefing section block (M9.3) — resolved from
    core_config via briefing_sections.resolve_briefing_sections(); when omitted
    or on any error it degrades to the Danny-era default (byte-identical).
    """
    from core.services.user_settings import resolve_user_name
    user_name = user_name or resolve_user_name()
    if sections is None:
        from core.services.briefing_sections import resolve_briefing_sections
        sections = resolve_briefing_sections()
    voice = get_voice()
    guards = inject_guards("briefing")
    # M9.4: the tenant's timezone label (settings → env → IST), so a
    # non-IST tenant's briefing is framed in THEIR time, not Danny's.
    from core.lib.time_utils import tz_label
    tz_lbl = tz_label()
    return f"""
ROLE: {voice}

You have full situational awareness of {user_name}'s {sections.role_framing}.
Your job is to give {user_name} a clear picture of the board so he can make his next move.

CURRENT TIME: {ctx.current_time_str}
CURRENT PHASE: {ctx.briefing_mode}
HEADLINE: Use exactly "{ctx.briefing_mode}" as your opening headline.
SYSTEM LOAD: {'OVERLOADED - urgent items need attention' if ctx.is_overloaded else 'STEADY'}
MONDAY REENTRY: {'TRUE - start with weekend recon' if ctx.is_monday_morning else 'FALSE'}
PEOPLE: {ctx.people_names}

--- OPENING (ALWAYS REQUIRED) ---
Every briefing MUST open with Rhodey's opening line before any section header. Never start with a section.
The opening is the headline line, then 1-2 sentences in Rhodey's voice that orient {user_name}: what's new, what's on top, what needs his attention.

COMPASS TONE (the opening is always written — only its content changes):
- HINDSIGHT_EMPTY: Open with the board itself — what's on top right now, what needs a decision. Never skip the opening.
- HINDSIGHT_STALE but not empty: Dry one-sentence acknowledgment, then the opening. (e.g. "Nothing new since this morning — the board hasn't moved.")
- Hindsight fresh: Weave insights into a forward-leaning opening.

PHASE FOCUS (used by system_persona above):
The system_persona line at the top already encodes the phase-specific focus. Do not override it here.

--- THE BOARD ---
Build these sections from the data below. Only include sections that have items.

{sections.board_lines}

--- SECTION RULES ---
1. DATA FIDELITY: Every task in {sections.fidelity_names} MUST appear verbatim in SYSTEM TASKS. Schedule from CALENDAR EVENTS. Hindsight is for opening synthesis only - never bullet points.
2. EMPTY SECTIONS: Omit any section with zero items. Never output "None today" or "Empty".
3. MAX 3 ITEMS per section. Append "...and X more in vault" if over.
4. BOLD revenue-critical tasks (payments, quotes, high-ticket items).
5. Commitments: Tasks marked [OWED TO: person] surface as "Owed to the client: contract". Tasks marked [WAITING ON: person] flag as blocked: "Waiting on the vendor for 6 days: contract".
6. The LINK RULE: If a task is derived from a URL in NEW INPUTS, embed the URL via Markdown: "ICON [Action] using [Source Title](URL)".
7. MONDAY RULE: If MONDAY REENTRY is TRUE, start with a "WEEKEND RECON" section summarizing weekend work ideas.
8. RECENCY BIAS: First sentence prioritizes NEW INPUTS. Use Master Pages for the "Why" behind the "What".
9. NO REPETITION: Never repeat identical phrasing (e.g. "100% bandwidth") in consecutive briefings.
10. WEEKEND FILTER: If weekend, do NOT suggest or list Work tasks.
11. NO task numbers, IDs, weights, scores, parentheses, or metadata in the output.
12. Never mention "Monday" unless it's actually the weekend.

--- MODE OVERRIDES ---
- URGENT mode: Hide {sections.urgent_hide}. Work and Done only.
- NIGHT mode: {sections.night_order}.

--- TONE AND STYLE ---
Tone: {voice} Direct, punchy, varied phrasing. Never use: {BLOCKED_WORDS}.
The banned list governs your own prose only — task titles quoted verbatim (DATA FIDELITY rule 1) always win over it.

Layout rules:
- Every section icon and every task MUST occupy its own individual line.
- Never combine tasks into a paragraph. Never put a paragraph between a section header and its task list.
- Every item must follow: "- ICON Task Title [Project]"
- Use actual newlines, not \n text. No markdown code blocks.

--- DATA CONTEXT ---
STRATEGIC CONTEXT: {ctx.season_config}
{ctx.session_memory_context}

CALENDAR EVENTS TODAY:
{ctx.calendar_context}

RECENT MEMORIES (semantically related):
{ctx.recent_memories_context if ctx.recent_memories_context else "None"}

HINDSIGHT:
{ctx.hindsight_context}

WEEKLY PATTERNS:
{ctx.weekly_patterns_str if ctx.weekly_patterns_str else "None"}

GRAPH INTELLIGENCE: {ctx.graph_task_context}

TASK DEPENDENCY MAP:
{ctx.dependency_context if ctx.dependency_context else "None"}

COMMUNICATION PATTERNS:
{ctx.social_graph_context if ctx.social_graph_context else "None"}

TEMPORAL INSIGHTS:
{ctx.temporal_context if ctx.temporal_context else "None"}

GRAPH CENTRALITY:
{ctx.centrality_context if ctx.centrality_context else "None"}

ADAPTIVE BRIEFING FEEDBACK:
{ctx.adaptive_context if ctx.adaptive_context else "None"}

MORNING PULSE GRAPH NARRATIVE:
{ctx.morning_pulse_narrative}

SERENDIPITY:
{ctx.serendipity_context if ctx.serendipity_context else "None"}

CANONICAL (Master Pages):
{ctx.canonical_context if ctx.canonical_context else "No Master Pages yet. Rely on raw context."}

CROSS-SYSTEM DELTA:
{ctx.delta_context if ctx.delta_context else "None"}

ACTIVE PRACTICES:
{ctx.practices_context if ctx.practices_context else "None"}

ACTIVE CLUSTERS:
{ctx.active_clusters_context}

ALL SYSTEM TASKS (for ID matching):
{ctx.universal_task_map}

ACTIVE TASKS (filtered by clusters + core projects):
{ctx.cluster_task_list}

TASKS AWAITING YOUR ATTENTION:
{ctx.urgency_lists}

RESOURCE PATTERNS (30-day window):
{ctx.pattern_context}

NEWLY ENRICHED RESOURCES:
{ctx.newly_enriched_context}

RECENTLY VAULTED URLs:
{ctx.recent_urls_context}

==============================
NEW INPUTS
==============================
{ctx.new_inputs}
==============================
NEW INPUT TAGS: {ctx.new_input_tags}

{guards}

--- TOP FOCAL ITEM SELECTION ---
Your JSON output includes a `top_focal_item` field that tells the app the
SINGLE most important thing {user_name} should focus on right now.

CRITICAL: Prefer ACTIVE TASKS over pending decisions. Only pick a pending
(graph_node / graph_edge) decision if there are ZERO actionable tasks.
{user_name} uses the Inbox for decisions — the focal card is for tasks first.

Pick ONE item from the data below. Only pick an item that {user_name} can actually
act on. Follow these rules:

1. ACTIONABLE ONLY: Never pick a task with direction="waiting_on" — {user_name}
   cannot act on blocked items. Never pick an item {user_name} has repeatedly
   dismissed (conversation history shows this).

2. IMPORTANCE OVER URGENCY: An overdue but trivial task (e.g. "Clean the
   garage") is LESS important than a strategic task with no deadline (e.g.
   "Meet the client on Phase 2"). Use your judgment, not just deadline.

3. INVISIBLE BLOCKERS: If {user_name} keeps ignoring an overdue task, it might be
   blocked or deprioritized — don't keep surfacing it. Pick something fresh.

4. REASON MATTERS: The `reason` field is shown to {user_name}. Make it specific:
   "The bank forms are blocking the fund transfer" NOT "This task is overdue."

5. ACTION LABEL BY TYPE: The `action_label` controls what the first button
   says. Set it based on the item type:
   - For "task":     "I'll do it"
   - For "graph_node": "Approve person"
   - For "graph_edge": "Review edge"
   - For other types: use a short verb ("View", "Create", "Review")

6. SET TO EMPTY if there's truly nothing worth surfacing (all is quiet).
   The app will show an "all clear" state instead.

Output format for top_focal_item:
```json
{{
  "type": "task",           // "task", "graph_node", "graph_edge", or other
  "item_id": "123",         // task ID or pending item ID from the data
  "title": "Fill the bank forms",
  "reason": "Blocking the fund transfer — the bank is waiting on these forms.",
  "urgency": "critical",    // "critical", "important", "normal"
  "action_label": "I'll do it"
}}
```
For a pending person node:
```json
{{
  "type": "graph_node",
  "item_id": "456",
  "title": "Approve: New person",
  "reason": "New person to add to your network",
  "urgency": "normal",
  "action_label": "Approve person"
}}
```
If nothing needs {user_name}'s attention, output an empty object {{}}.

--- HOME MODE SELECTION ---
Your JSON output includes a `home_mode` field that controls how the app's home screen
lays out information for {user_name}. Choose the mode that best fits the current context:

- "proceed" (default): Normal operations. Show Act cards with priority items.
  Use when there's a mix of tasks and decisions, and nothing is critical.

- "decide": {user_name} has pending decisions to make. Choose when:
  * There are 2+ pending graph nodes, edges, or channel items awaiting approval
  * The Inbox has items that need review
  * Decisions are the primary action item right now

- "sprint": Deep focus mode. Choose when:
  * 2+ tasks are urgent or overdue
  * SYSTEM LOAD is OVERLOADED
  * There's a clear priority that needs {user_name}'s full attention
  * A calendar event is coming up that requires preparation

- "catch_up": {user_name} has been away. Choose when:
  * Several new items appeared since the last briefing
  * CROSS-SYSTEM DELTA shows significant changes
  * It's the first briefing of the day (morning)
  * Tasks were completed since last check-in

- "wrap": End-of-day closure. Choose when:
  * It's evening (19:00+ {tz_lbl} / Intel phase)
  * {user_name} should transition from work to personal time
  * There are completed tasks to acknowledge
  * Focus should be on closing open loops

Pick the SINGLE best mode. Default to "proceed" if unsure.

NOTE: You are a briefing engine only. Your JSON output contains exactlyfour fields: `briefing`, `voice_line`, `home_mode`, and `top_focal_item`.
You do NOT create, complete, or modify any tasks, projects, people, resources, or clusters.
All task operations are handled by the Action Planner on the webhook path."""


def build_pulse_system_instruction(
    system_persona: str,
    briefing_history_context: str,
    routing_logic: str,
    drift_context: str = "None",
    user_name: str | None = None,
) -> str:
    """Build the pulse system instruction (M2 de-personalized).

    `user_name` comes from user_settings (fallback: env USER_NAME / "Danny").
    """
    from core.services.user_settings import resolve_user_name
    user_name = user_name or resolve_user_name()
    guards = inject_guards("briefing")
    # M9.4: the tenant's timezone label/offset (settings → env → IST) — the
    # HIGH-PRECISION TIME FORMATTING rule must match the tenant's zone.
    from core.lib.time_utils import tz_label, tz_offset_str
    tz_lbl = tz_label()
    tz_off = tz_offset_str()
    return f"""{system_persona}

    {briefing_history_context}

    MANDATE - SILENCE PROTOCOL & HALLUCINATION GUARD:
    - PROHIBIT ACTION HALLUCINATION: You are a logging tool, not an agent. NEVER say 'I'll ping', 'I'll check', 'I'll send', or 'I'll handle it'. You do not have the power to contact people. Your only job is to confirm that {user_name}'s task is SECURED in his system.
    - NEVER create a task from a URL unless {user_name} explicitly says "Make this a task."
    - NEVER proactively invent tasks or ideas. ONLY track what is manually entered or already exists.
    - NEVER "make up", guess, or generate example tasks.
    - NEVER mark an existing task as "done" unless NEW INPUTS explicitly contains a command matching that exact task.
    - ONLY track what is manually entered in NEW INPUTS.

    {routing_logic}

    DRIFT DETECTION (Temporal Lineage):
    - Check if active organizations have been updated 3+ times in 48 hours.
    - If DRIFT detected, add: "DRIFT ALERT: Organization '{{name}}' changed {{count}} times in 48h. Bottleneck?"
    - Use detect_drift(org_name) to check (returns update_count).

    SERENDIPITY PROTOCOL:
    - Under the "SERENDIPITY FINDS" context, you have been given a sample of multi-hop connections.
    - Review the connections. If you find a truly surprising, non-obvious link (e.g., a past meeting with someone related to today's task), mention it exactly as a one-sentence insight in the briefing.
    - STRICTLY FORBIDDEN: Do not merge multiple paths together. Do not hallucinate relationships. If all paths are boring, skip them entirely.

    STRATEGIC AUDIT INSTRUCTIONS:
    - BLINDSPOT AUDIT: Evaluate every URL in NEW INPUTS against {user_name}'s projects.
    - CONNECTION MAPPING: If a resource mentions a person in the PEOPLE list, link them in the summary.
    - PATTERN DETECTION: Review RECENTLY VAULTED RESOURCES and NEWLY ENRICHED RESOURCES. If you see 3+ related URLs on a new topic, mention the pattern in the briefing.
    - THE VAULT GATE: These observations are for the briefing only.
    - THE BRIEFING GATE: You are STRICTLY FORBIDDEN from mentioning new resources or new clusters in the briefing UNLESS {user_name} specifically used the word "Vault" or "Cluster" in the NEW INPUTS.

    CLUSTER vs. INCUBATOR FRAMEWORK:
    - CLUSTER ASSEMBLY: Evaluate every URL against ACTIVE CLUSTERS. If a URL provides a "component" for an existing cluster, mention this connection in the briefing.
    - THE INCUBATOR AUDIT: If an input represents a high-potential standalone product idea NOT related to current goals, flag it in the briefing.
    - SPARK DETECTION: If a link is a "Spark" (brand new project concept), note this in the briefing.

    DRIFT ALERTS (Temporal Lineage):
    {drift_context}

    INSTRUCTIONS:
    1. STRICT DATA FIDELITY: You are strictly forbidden from inventing or hallucinating data. Your single output is the `briefing` field. You do not create, complete, or modify any tasks or projects - the Action Planner handles all operations.
    2. ZERO-DUMP PROTOCOL: If NEW INPUTS is empty or "None", your briefing should simply report no new input. Do not generate empty sections.
    3. ANALYZE NEW INPUTS: Identify completions, new tasks, new people, and new projects for context - inform the briefing, do not action them.
    4. STRATEGIC NAG: If STAGNANT_URGENT_TASKS exists, start the brief by calling these out.
    5. STALE LOOPS: If STALE_TASKS exists, always include the Stale Loops section - never suppress it regardless of mode.
    6. HIGH-PRECISION TIME FORMATTING ({tz_lbl}/UTC{tz_off}): When {user_name} mentions a time, convert to ISO-8601. If DAY only (no time), output "YYYY-MM-DD". If EXACT TIME, output "YYYY-MM-DDTHH:MM:SS{tz_off}". NAKED TASKS: If NO date and NO time, return null for reminder_at.
    7a. RECURRENCE RULES: If {user_name} says "every Monday", "weekly", "daily", output an iCalendar RRULE string in "recurrence" (e.g., "RRULE:FREQ=WEEKLY;BYDAY=MO"). If he specifies an end date like "until December", append the UNTIL clause in UTC format (e.g., "RRULE:FREQ=WEEKLY;BYDAY=MO;UNTIL=20261231T000000Z"). Otherwise leave it null.
    8. STRATEGIC WEIGHTING: Highlight items based on the user's stated priorities in the briefing narrative.
    10. WEEKEND FILTER: If isWeekend is true, do NOT suggest or list Work tasks in the briefing.

    {guards}

    - Do not offer conclusions or summaries.
    - Maintain the stoic, concise voice of an AI assistant managing a heavy load.
    - If there's an active session memory, weave its context into the narrative.
"""

File: rhodey_app/lib/models/briefing.dart
// Models for the structured briefing response from GET /api/briefing.
//
// Sections: briefing (tasks + calendar), decisions (pending items), recent (outcomes),
// traces (paired input→outcome history for Traces view).
// Decisions section is omitted on the API when empty.

class BriefingItem {
  final String icon;   // "⏰", "✅", "🔗", "📝", "⚠️", etc.
  final String text;
  final String status; // "urgent", "active", "pending", "done", "note"

  /// Decision action fields — null for non-decision items.
  final String? decisionId;
  final String? decisionType;
  // "graph_node", "graph_edge", "email", "whatsapp", "call", "merge", "channel"

  /// Pulse intelligence: true if pulse flagged this as overdue/stale
  final bool isStale;

  const BriefingItem({
    required this.icon,
    required this.text,
    required this.status,
    this.decisionId,
    this.decisionType,
    this.isStale = false,
  });

  factory BriefingItem.fromJson(Map<String, dynamic> json) {
    return BriefingItem(
      icon: json['icon'] as String? ?? '📝',
      text: json['text'] as String? ?? '',
      status: json['status'] as String? ?? 'note',
      decisionId: json['decision_id'] as String?,
      decisionType: json['decision_type'] as String?,
      isStale: json['is_stale'] as bool? ?? false,
    );
  }

  bool get isUrgent => status == 'urgent';
  bool get isPending => status == 'pending';
  bool get isDecision => decisionId != null && decisionType != null;

  Map<String, dynamic> toJson() => {
        'icon': icon,
        'text': text,
        'status': status,
        'decision_id': decisionId,
        'decision_type': decisionType,
        'is_stale': isStale,
      };
}

class BriefingSection {
  final String id;    // "briefing", "decisions", "recent"
  final String title; // "Your morning", "Decisions", "Recent"
  final List<BriefingItem> items;

  const BriefingSection({
    required this.id,
    required this.title,
    required this.items,
  });

  factory BriefingSection.fromJson(Map<String, dynamic> json) {
    final rawItems = json['items'] as List<dynamic>? ?? [];
    return BriefingSection(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      items: rawItems
          .map((e) => BriefingItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'items': items.map((e) => e.toJson()).toList(),
      };
}

/// A paired input→outcome trace for the Traces view.
class TraceItem {
  /// Human-readable time: "2m ago", "1h ago"
  final String time;

  /// What the user said/asked (brief)
  final String input;

  /// What happened / outcome
  final String resolution;

  const TraceItem({
    required this.time,
    required this.input,
    required this.resolution,
  });

  factory TraceItem.fromJson(Map<String, dynamic> json) {
    return TraceItem(
      time: json['time'] as String? ?? '',
      input: json['input'] as String? ?? '',
      resolution: json['resolution'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'time': time,
        'input': input,
        'resolution': resolution,
      };
}

/// A catch-up delta item — what changed since the user was last active.
class DeltaItem {
  final String icon;   // "🆕", "✅", "🔗", "👤"
  final String text;
  final String time;   // "2m ago", "1h ago"

  const DeltaItem({
    required this.icon,
    required this.text,
    required this.time,
  });

  factory DeltaItem.fromJson(Map<String, dynamic> json) {
    return DeltaItem(
      icon: json['icon'] as String? ?? '📝',
      text: json['text'] as String? ?? '',
      time: json['time'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'icon': icon,
        'text': text,
        'time': time,
      };
}

/// A wrap item — completed today or rolling to tomorrow.
class WrapItem {
  final String icon;   // "✅", "📋"
  final String text;   // Task title
  final String detail; // Detail line: time ago, priority, etc.

  const WrapItem({
    required this.icon,
    required this.text,
    required this.detail,
  });

  factory WrapItem.fromJson(Map<String, dynamic> json) {
    return WrapItem(
      icon: json['icon'] as String? ?? '📝',
      text: json['text'] as String? ?? '',
      detail: json['detail'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'icon': icon,
        'text': text,
        'detail': detail,
      };
}

class BriefingResponse {
  final String greeting;
  final String? nextEvent;
  final List<BriefingSection> sections;
  final int pendingCount;
  final List<TraceItem> traces;
  final String? latestResponse; // Most recent bot response text

  // Pulse intelligence fields
  final String? contextBar;      // "Closing the loop — clear banking before sign-off"
  final String? voiceLine;       // Rhodey's voice line (1-2 sentences)
  final String? pulseMode;       // "morning", "afternoon", "weekend", etc.
  final List<String> insights;   // ["🔴 2 stale tasks", "📦 3 vaulted"]
  final int vaultedCount;        // Tasks hidden by pulse vault

  // Home screen mode (drives Flutter layout)
  final String homeMode;         // "proceed" | "decide" | "sprint" | "catch_up" | "wrap"
  final int vaultedUrgentCount;  // Vaulted urgent tasks (shown as 🔴 count)
  final int vaultedHighCount;    // Vaulted high-priority tasks (shown as 🟡 count)
  // Catch-up delta items
  final List<DeltaItem> deltaItems; // What changed since last session
  // Wrap mode: completed today + rolling tasks
  final List<WrapItem> wrapDoneToday;
  final List<WrapItem> wrapRolling;

  const BriefingResponse({
    required this.greeting,
    this.nextEvent,
    required this.sections,
    this.pendingCount = 0,
    this.traces = const [],
    this.latestResponse,
    this.contextBar,
    this.voiceLine,
    this.pulseMode,
    this.insights = const [],
    this.vaultedCount = 0,
    this.homeMode = 'proceed',
    this.vaultedUrgentCount = 0,
    this.vaultedHighCount = 0,
    this.deltaItems = const [],
    this.wrapDoneToday = const [],
    this.wrapRolling = const [],
  });

  factory BriefingResponse.fromJson(Map<String, dynamic> json) {
    final rawSections = json['sections'] as List<dynamic>? ?? [];
    final rawTraces = json['traces'] as List<dynamic>? ?? [];
    final rawInsights = json['insights'] as List<dynamic>? ?? [];
    final rawDelta = json['delta_items'] as List<dynamic>? ?? [];
    final rawDoneToday = json['wrap_done_today'] as List<dynamic>? ?? [];
    final rawRolling = json['wrap_rolling'] as List<dynamic>? ?? [];
    return BriefingResponse(
      greeting: json['greeting'] as String? ?? 'Hey.',
      nextEvent: json['next_event'] as String?,
      sections: rawSections
          .map((e) => BriefingSection.fromJson(e as Map<String, dynamic>))
          .toList(),
      pendingCount: json['pending_count'] as int? ?? 0,
      traces: rawTraces
          .map((e) => TraceItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      latestResponse: json['latest_response'] as String?,
      contextBar: json['context_bar'] as String?,
      voiceLine: json['voice_line'] as String?,
      pulseMode: json['pulse_mode'] as String?,
      insights: rawInsights.whereType<String>().toList(),
      vaultedCount: json['vaulted_count'] as int? ?? 0,
      homeMode: json['home_mode'] as String? ?? 'proceed',
      vaultedUrgentCount: json['vaulted_urgent_count'] as int? ?? 0,
      vaultedHighCount: json['vaulted_high_count'] as int? ?? 0,
      deltaItems: rawDelta
          .map((e) => DeltaItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      wrapDoneToday: rawDoneToday
          .map((e) => WrapItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      wrapRolling: rawRolling
          .map((e) => WrapItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  /// Empty briefing (e.g. on error or initial load)
  static BriefingResponse empty() => const BriefingResponse(
        greeting: 'Hey.',
        sections: [],
      );

  Map<String, dynamic> toJson() => {
        'greeting': greeting,
        'next_event': nextEvent,
        'sections': sections.map((e) => e.toJson()).toList(),
        'pending_count': pendingCount,
        'traces': traces.map((e) => e.toJson()).toList(),
        'latest_response': latestResponse,
        'context_bar': contextBar,
        'voice_line': voiceLine,
        'pulse_mode': pulseMode,
        'insights': insights,
        'vaulted_count': vaultedCount,
        'home_mode': homeMode,
        'vaulted_urgent_count': vaultedUrgentCount,
        'vaulted_high_count': vaultedHighCount,
        'delta_items': deltaItems.map((e) => e.toJson()).toList(),
        'wrap_done_today': wrapDoneToday.map((e) => e.toJson()).toList(),
        'wrap_rolling': wrapRolling.map((e) => e.toJson()).toList(),
      };
}
